File: Cliente Python/Cliente.py
import socket   # Libreria para poder usar sockets
import os       # Libreria para usar funciones del sistema operativo

socketCliente = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # Crea un socket

# Dirección IP y puerto del servidor
servidor_ip = '127.0.0.1'  # localhost
servidor_puerto = 3000

socketCliente.connect((servidor_ip, servidor_puerto)) # El cliente se conecta al servidor
print('Conectado al servidor') # Escribe un mensaje en la consola

while(True):

    opcion = 0

    while(True):

        print('Indique que desea generar'); 
        print('1) Nombre de usuario')
        print('2) Contraseña')
        print('Ingrese 0 para salir')

        while(True):
            opcion = input() # El usuario ingresa algo. Todo lo que ingrese se considera como un string
            
            if(opcion.isdigit() ):
                opcion = int(opcion) # Transformo el string a int
                break
            else:
                print('Ingrese solo un número')

        if(opcion == 1 or opcion == 2 or opcion == 0): break
        else:                                          print("Elija una opción valida")

    os.system('cls') # Limpia los mensajes de la consola



    if(opcion == 1):
        
        # Informo al servidor de la elección del usuario
        mensaje = '1\0'                          # Como el servidor esta en C, le agrego el \0 al final del mensaje porque los strings en C terminan con \0. Caso contrario, el servidor lee basura
        socketCliente.sendall(mensaje.encode() ) # Codifica un mensaje y lo envia al servidor
        
        while True:
            print('Indique cuantos caracteres quiere en su nombre de usuario (debe ser entre 5 y 15)')
            tamanio = 0

            while(True):
                tamanio = input() 
                
                if(tamanio.isdigit() ): break
                else:                   print('Ingrese solo un número')
            
            mensaje = tamanio + '\0'
            socketCliente.sendall(mensaje.encode() )

            respuesta = socketCliente.recv(1024).decode('utf-8').strip() # Recibe un mensaje del servidor. También lo decodifica de binario a string y le borra los espacios vacios
            respuesta = respuesta[:-1] # Elimino el último caracter del string. Los mensajes que vienen desde el servidor son strings C, es decir, siempre terminan con \0 al final

            if(respuesta == 'Longitud valida'):
                nombreGenerado = socketCliente.recv(1024).decode('utf-8')
                print('El nombre de usuario generado es: ' + nombreGenerado)
                break
            else:
                print(respuesta)
        


    if(opcion == 2):

        mensaje = '2\0'                          
        socketCliente.sendall(mensaje.encode() )

        while True:
            print('Indique cuantos caracteres quiere en su nombre de usuario (debe ser entre 8 y 50)')
            tamanio = 0

            while(True):
                tamanio = input() 
                
                if(tamanio.isdigit() ): break
                else:                   print('Ingrese solo un número')
            
            mensaje = tamanio + '\0'
            socketCliente.sendall(mensaje.encode() )

            respuesta = socketCliente.recv(1024).decode('utf-8').strip()
            respuesta = respuesta[:-1] 

            if(respuesta == 'Longitud valida'):
                codigoGenerado = socketCliente.recv(1024).decode('utf-8')
                print('El código generado es: ' + codigoGenerado)
                break
            else:
                print(respuesta)



    if(opcion == 0):

        mensaje = '0\0'                          
        socketCliente.sendall(mensaje.encode() )
        socketCliente.close() # Cierra el socket
        break


    print("Presiona Enter para continuar...")
    input()
    os.system('cls')

File: Cliente C/Cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h> // Permite usar booleanos
#include <string.h>

#include <locale.h>   // Libreria para localizar el lenguaje de los caracteres
#include <winsock2.h> // Librería para la programar sockets en Windows

#include "Cliente.h"

/************************************* FUNCIONES DEL SOCKET ******************************************************/

SOCKET conectarseAlServidor()
{
    setlocale(LC_ALL, ""); // Establezco que el programa maneje caracteres especiales


    /* INICIALIZAR WINSOCK */
    WSADATA wsaData; // Creo una estructura WSADATA para implementar Winsock(Windows Sockets)

    // WSAStartup() inicia el uso de winsock. MAKEWORD(2, 2) indica que se usa la versión 2.2 de Winsock
    if( WSAStartup(MAKEWORD(2,2), &wsaData) != 0 )
    {
        printf("Error al inicializar Winsock \n");
        return 1;
    }


    /* CREAR EL SOCKET */
    SOCKET socketCliente = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);  // La función socket() crea un socket. Sus parámetros son dominio, tipo de socket y protocolo

    //  Los dominios pueden ser:                    Los tipos pueden ser:                       Los protocolos pueden ser:
    //    AF_INET para IPv4                             SOCK_STREAM para una conexión TCP           IPPROTO_TCP para el protocolo TCP
    //    AF_INET6 para IPv6                            SOCK_DGRAM para una conexión UDP
    //    AF_UNIX para comunicación local

    if (socketCliente == INVALID_SOCKET)
    {
        printf("Error al crear el socket \n");
        WSACleanup(); // WSACleanup() finaliza el uso de Winsock
        return 1;
    }


    /* CONFIGURAR LA DIRECCIÓN Y PUERTO */
    SOCKADDR_IN servidorAddress; // SOCKADDR_IN es una estructura para manejar las direcciones del servidor socket

    servidorAddress.sin_family = AF_INET;          // sin_family es la familia de direcciones
    servidorAddress.sin_port = htons(3000);        // sin_port es el puerto. La función htons (host to network short) convierte el número de puerto del orden de bytes del host al orden de bytes de la red
    servidorAddress.sin_addr.s_addr =  inet_addr("127.0.0.1");  // sin_addr.s_addr es la dirección IP del servidor. En INADDR_ANY significa que aceptará conexiones en cualquiera de las interfaces de red disponibles en la máquina.


    /* CONECTARSE CON EL SERVIDOR */
    bool conexionExitosa = false;
    while(!conexionExitosa) // Se queda en un loop infinito hasta que se conecte
    {
        if (connect(socketCliente, (struct servidorAddress*)&servidorAddress, sizeof(servidorAddress)) == SOCKET_ERROR)
        {
            //system("cls"); // Limpia la pantalla. Prefiero que no lo haga para que así se vea que de verdad esta intentando conectarse de vuelta
            printf("Error al conectar con el servidor. Intentando de vuelta \n");
        }
        else
        {
            printf("Conectado con el servidor \n");
            conexionExitosa = true;
        }
    }

    return socketCliente;
}


void enviarMensaje(char *mensaje, SOCKET socketCliente)
{
    send(socketCliente, mensaje, strlen(mensaje)+1, 0); // send() envia un mensaje. Sus parámetros son el socket a través del cual se enviarán los datos, el buffer y el tamaño en bytes de los datos que se desean enviar
}


/************************************* MENU ******************************************************/

int menu()
{
    int opcion;
    boolean opcionValida = false;

    while(!opcionValida)
    {
        printf("Indique que desea generar \n");
        printf("1) Nombre de usuario \n");
        printf("2) Contraseña \n");
        printf("Ingrese 0 para salir \n");

        opcion = 0;

        while(true) // Me aseguro que el cliente solo ponga un int
        {
            if(scanf("%d", &opcion) )
            {
                fflush(stdin);
                break;
            }
            else
            {
                fflush(stdin);
                printf("Ingrese solo un número \n");
            }
        }

        if(opcion == 1 || opcion == 2 || opcion == 0) opcionValida = true;
        else                                          printf("Elija una opción valida \n");
    }

    system("cls");
    return opcion;
}


/************************************************** PUNTO 1.A ***********************************************************/

void generarNombreDeUsuario(SOCKET socketCliente)
{
    char buffer[1024];

    while(true)
    {
        printf("Indique cuantos caracteres quiere en su nombre de usuario (debe ser entre 5 y 15) \n");
        int tamanio = ingresarTamanio();

        char mensaje[sizeof(tamanio)];
        sprintf(mensaje, "%d", tamanio);    // Convierto el int a string
        enviarMensaje(mensaje, socketCliente);

        recv(socketCliente, buffer, sizeof(buffer), 0);

        if(strcmp(buffer, "Longitud valida") == 0)      // Si el servidor no me da el visto bueno, se repite todo
        {
            memset(buffer, '\0', strlen(buffer) );

            recv(socketCliente, buffer, sizeof(buffer), 0);
            printf("El nombre de usuario generado es: %s \n", buffer);
            memset(buffer, '\0', strlen(buffer) );

            system("pause");
            system("cls");
            break;
        }
        else
        {
            printf("%s\n", buffer);
            memset(buffer, '\0', strlen(buffer) );
        }
    }

}

/************************************************** PUNTO 1.B ***********************************************************/
void generarContrasenia(SOCKET socketCliente)
{
    char buffer[1024];

    while(true)
    {
        printf("Indique cuantos caracteres quiere en código (debe ser entre 8 y 50) \n");
        int tamanio = ingresarTamanio();

        char mensaje[sizeof(tamanio)];
        sprintf(mensaje, "%d", tamanio);    // Convierto el int a string
        enviarMensaje(mensaje, socketCliente);

        recv(socketCliente, buffer, sizeof(buffer), 0);

        if(strcmp(buffer, "Longitud valida") == 0)      // Si el servidor no me da el visto bueno, se repite todo
        {
            memset(buffer, '\0', strlen(buffer) );

            recv(socketCliente, buffer, sizeof(buffer), 0);
            printf("El código generado es: %s \n", buffer);
            memset(buffer, '\0', strlen(buffer) );

            system("pause");
            system("cls");
            break;
        }
        else
        {
            printf("%s\n", buffer);
            memset(buffer, '\0', strlen(buffer) );
        }
    }

}


/****************************************** UTIL *****************************************************/
int ingresarTamanio()
{
    int tamanio;

    while(true) // Me aseguro que el cliente solo ponga un int
    {
        if(scanf("%d", &tamanio) )
        {
            fflush(stdin);
            break;
        }
        else
        {
            fflush(stdin);
            printf("Ingrese solo un número \n");
        }
    }

    return tamanio;
}
